### PR TITLE
basics hometask done

### PR DESCRIPTION
**Pull request by kusem**

### CODE REVIEW

File: src/basics/BasicsValidator.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
+<?php
+
```
Comment (yderevianko-cartera) [inline src/basics/BasicsValidator.php:2]: not needed empty lines

File: src/basics/BasicsValidator.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace basics;
+
+
```
Comment (yderevianko-cartera) [inline src/basics/BasicsValidator.php:6]: not needed empty line

File: src/basics/BasicsValidator.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+use InvalidArgumentException;
+
+class BasicsValidator implements BasicsValidatorInterface
+{
+
```
Comment (yderevianko-cartera) [inline src/basics/BasicsValidator.php:11]: not needed empty line

File: src/basics/BasicsValidator.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    /**
+     * @inheritDoc
+     */
+    public function isValidStringException(string $input): void
+    {
+        if (strlen($input) != 6) {
+            throw new InvalidArgumentException('String should have length=6');
+        }
+    }
+}
```
Comment (yderevianko-cartera) [inline src/basics/BasicsValidator.php:41]: please add empty line at the end of file

File: src/basics/basics.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
+<?php
+
```
Comment (yderevianko-cartera) [inline src/basics/basics.php:2]: not needed empty line

File: src/basics/basics.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+namespace basics;
+
+
```
Comment (yderevianko-cartera) [inline src/basics/basics.php:6]: not needed empty line

File: src/basics/basics.php [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+use JetBrains\PhpStorm\Pure;
+
+class basics implements BasicsInterface
+{
+    private BasicsValidator $validator;
+
+    #[Pure] public function __construct()
```
Comment (yderevianko-cartera) [inline src/basics/basics.php:13]: please remove PHPStorm annotations

File: src/basics/basics.php [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+            return "fourth";
+        }
+        if ($minute <= 15) {
+            return "first";
+        }
+
+        if ($minute <= 30) {
+            return "second";
+        }
+        return "third";
```
Comment (yderevianko-cartera) [inline src/basics/basics.php:35]: please add empty line before return

File: src/basics/basics.php [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        $this->validator->isYearException($year);
+
+        return ((($year % 4) == 0) && ((($year % 100) != 0) || (($year % 400) == 0)));
+    }
+
+    /**
+     * @inheritDoc
+     */
+
```
Comment (yderevianko-cartera) [inline src/basics/basics.php:51]: please remove empty line

File: src/basics/basics.php [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        $convertedString = str_split($input);
+        $firstHalf =
+            (int)$convertedString[0] +
+            (int)$convertedString[1] +
+            (int)$convertedString[2];
+        $secondHalf =
+            (int)$convertedString[3] +
+            (int)$convertedString[4] +
+            (int)$convertedString[5];
+        return $firstHalf == $secondHalf;
```
Comment (yderevianko-cartera) [inline src/basics/basics.php:65]: please add empty line before return

File: src/basics/basics.php [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getMinuteQuarter(int $minute): string
+    {
+        $this->validator->isMinutesException($minute);
+
+        if ($minute > 45 or $minute == 0) {
```
Comment (yderevianko-cartera) [inline src/basics/basics.php:25]: Please read difference between 'or' and '||' and use || and && instaed of 'or' 'and'

File: src/basics/basics.php [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+            (int)$convertedString[0] +
+            (int)$convertedString[1] +
+            (int)$convertedString[2];
+        $secondHalf =
+            (int)$convertedString[3] +
+            (int)$convertedString[4] +
+            (int)$convertedString[5];
+        return $firstHalf == $secondHalf;
+    }
+}
```
Comment (yderevianko-cartera) [inline src/basics/basics.php:67]: please add empty line at the end of file

**Review by yderevianko-cartera — CHANGES_REQUESTED** [review]

File: src/basics/BasicsValidator.php (hunk 1 above)
Comment (kusem) [reply src/basics/BasicsValidator.php:2]: > not needed empty lines

but PSR-12 requires empty line there...
And phpstorm automatically adds it while auto code-reformat

**Comment by yderevianko-cartera** [comment]
@kusem please do not close pull request, just commit and push changes and it will automatically update.

**Comment by kusem** [comment]
> @kusem please do not close pull request, just commit and push changes and it will automatically update.

Sorry, just found that I've did a very big mistake - every next home task was made using all ready previous home tasks, so each PR had the same errors. Just fixing it now by creating new PRs with separated branches.

**Pull request closed without merging**